libsupport/include/katana/HostAllocator.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <utility>
+
+#include "katana/config.h"
+
+namespace katana {
+
+enum class MemoryPinType { Swappable = 0, Pinned = 1 };
+
+class KATANA_EXPORT HostHeap {
+public:
+  //! Allocate uninitialized items
+  virtual void* Malloc(const size_t n_bytes) = 0;
+  //! Allocate bitwise-zero-initialized items
+  virtual void* Calloc(const size_t n_items, const size_t item_size) = 0;
+  //! Reallocate the buffer of new size, copying the old data. O(N+M) space, where N is the old size and M is the new size.
+  virtual void* Realloc(void* ptr, const size_t new_bytes) = 0;
+  //! Release memory
+  virtual void Free(void* ptr) = 0;
+  //! Return the runtime memory pin type used by the hh_
+  virtual MemoryPinType GetType() const = 0;
+
+  virtual ~HostHeap();
+};
+
+class KATANA_EXPORT SwappableHostHeap : public HostHeap {
+public:
+  void* Malloc(const size_t n_bytes) override { return malloc(n_bytes); }
+
+  void* Calloc(const size_t n_items, size_t item_size) override {
+    return calloc(n_items, item_size);
+  }
+
+  void* Realloc(void* ptr, const size_t new_bytes) override {
+    return realloc(ptr, new_bytes);
+  }
+
+  void Free(void* ptr) override { free(ptr); }
+
+  MemoryPinType GetType() const override { return MemoryPinType::Swappable; }
+
+  ~SwappableHostHeap() override;
+};
+
+KATANA_EXPORT SwappableHostHeap* GetSwappableHostHeap();
+
+template <typename Ty>
+class HostAllocator {
+  HostHeap* hh_;
+
+  inline void destruct(char*) const {}
+  inline void destruct(wchar_t*) const {}
+  template <typename T>
+  inline void destruct(T* t) const {
```

Comment:
We can use `enable_if` trick to enable destruction only for class types. Something like:
```
template <typename T>
std::enable_if_t<!std::is_scalar_v<T>> destruct(T* t) const {
   t->~T();
}

template <typename T>
std::enable_if_t<std::is_scalar_v<T>> destruct(T* t) const {}
```

Reply:
Thanks for the exact code, it works.